contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,435 @@
+        require(_proof.length == PROOF_LENGTH + 0x20*5, "wrong proof length");
+        uint256[2] memory publicKey;
+        uint256 preSeed;
+        uint256 blockNum;
+        address sender;
+        assembly { // solhint-disable-line no-inline-assembly
+            publicKey := add(_proof, PUBLIC_KEY_OFFSET)
+            preSeed := mload(add(_proof, PRESEED_OFFSET))
+            blockNum := mload(add(_proof, BLOCKNUM_OFFSET))
+            // We use a struct to limit local variables to avoid stack depth errors.
+            mstore(fp, mload(add(add(_proof, BLOCKNUM_OFFSET), 0x20)))
+            mstore(add(fp, 0x20), mload(add(add(_proof, BLOCKNUM_OFFSET), 0x40)))
+            mstore(add(fp, 0x40), mload(add(add(_proof, BLOCKNUM_OFFSET), 0x60)))
+            sender := mload(add(add(_proof, BLOCKNUM_OFFSET), 0x80))
+        }
+        currentKeyHash = hashOfKey(publicKey);
+        bytes32 callback = s_callbacks[preSeed];
+        requestId = preSeed;
+        require(callback != 0, "no corresponding request");
+        require(callback == keccak256(abi.encodePacked(requestId, blockNum, fp.subId, fp.callbackGasLimit, fp.numWords, sender)), "incorrect commitment");
+        fp.sender = sender;
+
+        bytes32 blockHash = blockhash(blockNum);
+        if (blockHash == bytes32(0)) {
+            blockHash = BLOCKHASH_STORE.getBlockhash(blockNum);
+            require(blockHash != bytes32(0), "please prove blockhash");
+        }
+        // The seed actually used by the VRF machinery, mixing in the blockhash
+        uint256 actualSeed = uint256(keccak256(abi.encodePacked(preSeed, blockHash)));
+        // solhint-disable-next-line no-inline-assembly
+        assembly { // Construct the actual proof from the remains of _proof
+            mstore(add(_proof, PRESEED_OFFSET), actualSeed)
+            mstore(_proof, PROOF_LENGTH)
+        }
+        randomness = VRF.randomValueFromVRFProof(_proof); // Reverts on failure
+    }
+
+    function getFeedData()
+    private
+    view
+    returns (
+        uint256,
+        uint256
+    )
+    {
+        uint32 stalenessSeconds = s_config.stalenessSeconds;
+        bool staleFallback = stalenessSeconds > 0;
+        uint256 timestamp;
+        int256 gasWei;
+        int256 linkEth;
+        // Fallback to the fallback price if the feed is too stale.
+        // Maybe need to optimize to avoid this contract call and used a cached
+        // price, say in the case of a large number of fulfillments in short succession?
+        (,gasWei,,timestamp,) = FAST_GAS_FEED.latestRoundData();
+        if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
+            gasWei = s_fallbackGasPrice;
+        }
+        (,linkEth,,timestamp,) = LINK_ETH_FEED.latestRoundData();
+        if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
+            linkEth = s_fallbackLinkPrice;
+        }
+        return (uint256(gasWei), uint256(linkEth));
```

Comment:
In an updated version of Keeper we have added a check to make sure that these are greater than -1 before casting. The feeds should never return that value, but the result of either being negative would be a big enough problem that we guard against it.